android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/ConfigurationCommandCreator.kt

```diff
@@ -0,0 +1,205 @@
+import com.badoo.ribs.core.routing.configuration.feature.BackStackElement
+import com.badoo.ribs.core.routing.configuration.feature.BackStackFeature
+import io.reactivex.Observable
+import java.lang.Math.min
+
+/**
+ * Takes the state emissions from [BackStackFeature], and translates them to a stream of
+ * [ConfigurationCommand]s.
+ *
+ * @see [ConfigurationCommandCreator.diff]
+ */
+internal fun <C : Parcelable> BackStackFeature<C>.commands(): Observable<ConfigurationCommand<C>> =
+    Observable.wrap(this)
+        .startWith(initialState) // Bootstrapper can overwrite it by the time we receive the first state emission here
+        .buffer(2, 1)
+        .map { ConfigurationCommandCreator.diff(it[0].backStack, it[1].backStack) }
+        .flatMapIterable { items -> items }
+
+internal object ConfigurationCommandCreator {
+
+    /**
+     * Calculates diff between two states of [BackStackFeature] (where each state contains the current list
+     * of [C] configurations representing the back stack), and translates the diff to
+     * a list of [ConfigurationCommand]s that represent getting from the old state to the new one.
+     */
+    fun <C : Parcelable> diff(
+        oldStack: List<BackStackElement<C>>,
+        newStack: List<BackStackElement<C>>
+    ): List<ConfigurationCommand<C>> =
+        when (newStack) {
+            oldStack -> emptyList()
+            else -> {
+                val commands = mutableListOf<ConfigurationCommand<C>>()
+                val indexOfLastCommonElement = findIndexOfLastCommonElement(oldStack, newStack)
+                commands += oldStack.deactivateIfNeeded(newStack)
+                commands += oldStack.removeUntil(indexOfLastCommonElement)
+                commands += newStack.addFrom(indexOfLastCommonElement)
+                commands += newStack.activateIfNeeded(oldStack)
+                commands += newStack.overlayDiff(oldStack)
+                commands
+        }
+    }
+
+    private fun findIndexOfLastCommonElement(
+        oldStack: List<BackStackElement<*>>,
+        newStack: List<BackStackElement<*>>
+    ): Int {
+        var idx = -1
+
+        for (i in 0..min(oldStack.lastIndex, newStack.lastIndex)) {
+            // Compare only configurations, as if it's only the overlays that are different, they will be handled later
+            if (newStack.elementAt(i).configuration != oldStack.elementAt(i).configuration) {
+                return idx
+            }
+
+            idx = i
+        }
+
+        return idx
+    }
+
+    private fun <C : Parcelable> List<BackStackElement<C>>.deactivateIfNeeded(other: List<BackStackElement<C>>): List<ConfigurationCommand<C>> =
+        when {
+            contentListHasChanged(other) && isNotEmpty() -> {
+                val contentKey = Content(lastIndex)
+                val commands = mutableListOf<ConfigurationCommand<C>>()
+                commands += last().deactivateAllOverlays(contentKey)
```

Comment:
listOf(
   ...
).flatten()

Reply:
I'd rather keep the current format everywhere across the class, seems more readable.